engine/src/main/java/org/pentaho/di/trans/TransProfileFactory.java

```diff
@@ -0,0 +1,232 @@
+      new int[] { GroupByMeta.TYPE_GROUP_MIN, GroupByMeta.TYPE_GROUP_MAX, GroupByMeta.TYPE_GROUP_COUNT_ALL, };
+
+    int[] dateCalculations =
+      new int[] { GroupByMeta.TYPE_GROUP_MIN, GroupByMeta.TYPE_GROUP_MAX, GroupByMeta.TYPE_GROUP_COUNT_ALL, };
+
+    int[] booleanCalculations =
+      new int[] { GroupByMeta.TYPE_GROUP_MIN, GroupByMeta.TYPE_GROUP_MAX, GroupByMeta.TYPE_GROUP_COUNT_ALL, };
+
+    // Run it through the "group by" step without a grouping.
+    // Later, we can use the UnivariateStats plugin/step perhaps.
+    //
+    GroupByMeta statsMeta = new GroupByMeta();
+    int nrNumeric = 0;
+    int nrDates = 0;
+    int nrStrings = 0;
+    int nrBooleans = 0;
+    for ( ValueMetaInterface valueMeta : tableLayout.getValueMetaList() ) {
+      if ( valueMeta.isNumeric() ) {
+        nrNumeric++;
+      }
+      if ( valueMeta.isDate() ) {
+        nrDates++;
+      }
+      if ( valueMeta.isString() ) {
+        nrStrings++;
+      }
+      if ( valueMeta.isBoolean() ) {
+        nrBooleans++;
+      }
+    }
+    int nrCalculations =
+      nrNumeric
+        * numericCalculations.length + nrDates * dateCalculations.length + nrStrings
+        * stringCalculations.length + nrBooleans * booleanCalculations.length;
+
+    statsMeta.allocate( 0, nrCalculations );
+    int calcIndex = 0;
+    for ( int i = 0; i < tableLayout.size(); i++ ) {
+      ValueMetaInterface valueMeta = tableLayout.getValueMeta( i );
+      // Numeric data...
+      //
+      if ( valueMeta.isNumeric() ) {
+        //CHECKSTYLE:Indentation:OFF
+        //CHECKSTYLE:LineLength:OFF
+        for ( int c = 0; c < numericCalculations.length; c++ ) {
+          statsMeta.getAggregateField()[calcIndex] = valueMeta.getName() + "(" + GroupByMeta.getTypeDesc( numericCalculations[c] ) + ")";
+          statsMeta.getSubjectField()[calcIndex] = valueMeta.getName();
+          statsMeta.getAggregateType()[calcIndex] = numericCalculations[c];
+          calcIndex++;
+        }
+      }
+
+      // String data
+      //
+      if ( valueMeta.isString() ) {
+        //CHECKSTYLE:Indentation:OFF
+        //CHECKSTYLE:LineLength:OFF
+        for ( int c = 0; c < stringCalculations.length; c++ ) {
+          statsMeta.getAggregateField()[calcIndex] = valueMeta.getName() + "(" + GroupByMeta.getTypeDesc( stringCalculations[c] ) + ")";
+          statsMeta.getSubjectField()[calcIndex] = valueMeta.getName();
+          statsMeta.getAggregateType()[calcIndex] = stringCalculations[c];
+          calcIndex++;
+        }
+      }
+
+      // Date data
+      //
+      if ( valueMeta.isDate() ) {
+        for ( int c = 0; c < dateCalculations.length; c++ ) {
+          statsMeta.getAggregateField()[calcIndex] =
+            valueMeta.getName() + "(" + GroupByMeta.getTypeDesc( dateCalculations[c] ) + ")";
+          statsMeta.getSubjectField()[calcIndex] = valueMeta.getName();
+          statsMeta.getAggregateType()[calcIndex] = dateCalculations[c];
+          calcIndex++;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 173:179, 185:191.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.